### PR TITLE
AppDeployment: Add definitions and utilities code

### PR DESCRIPTION
**Pull request by hongchaodeng**
This takes out all the definitions and utilities code from https://github.com/oam-dev/kubevela/pull/1193 .
The controller code needs to be rewritten so let's get this part in first.

### CODE REVIEW

File: apis/core.oam.dev/v1alpha2/appdeployment_types.go [hunk 1]
```diff
@@ -0,0 +1,197 @@
+/*
+Copyright 2020 The KubeVela Authors.
```
Comment (wonderflow) [inline apis/core.oam.dev/v1alpha2/appdeployment_types.go:2]:
```suggestion
Copyright 2021 The KubeVela Authors.
```

**Review by wonderflow — CHANGES_REQUESTED** [review]
I didn't see the added imports in `go.mod` used any where?

**Comment by hongchaodeng** [comment]
> I didn't see the added imports in go.mod used any where?

@wonderflow 

in common.Scheme

**Comment by hongchaodeng** [comment]
@wonderflow 

No everything in go.mod is used because some of them are indirect imports.

**Comment by hongchaodeng** [comment]
@wonderflow 

They are automatically imported via `go get` command. Let me give you one example, if we removed `github.com/go-logr/zapr`, then we will have the following error:

```
# github.com/go-logr/zapr
../../../../pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:136:10: cannot use &infoLogger{...} (type *infoLogger) as type logr.Logger in return argument:
	*infoLogger does not implement logr.Logger (missing Error method)
../../../../pkg/mod/github.com/go-logr/zapr@v0.1.0/zapr.go:141:2: cannot use disabledInfoLogger (type *noopInfoLogger) as type logr.Logger in return argument:
	*noopInfoLogger does not implement logr.Logger (missing Error method)
```

**Review by wonderflow — APPROVED** [review]
LGTM after CI passed

**Pull request merged**